### PR TITLE
Sort files with C locale

### PR DESCRIPTION
**Pull request by jennybc**
I think `decor::cpp_files()` should return a file listing according to an explicit fixed locale and C is the obvious choice. In particular, this eliminates nuisance churn in `R/cpp11.R` and `src/cpp11.cpp`, which are created by `cpp11::cpp_register()`. Below I describe how to reproduce the churn. The culprit in my current usage is the RStudio IDE.

There are a few ways to get the C locale ordering:

* Use `withr::with_locale()`. But withr isn't already a direct or indirect dependency, so I didn't do this.
* Inline an equivalent helper, like this one from roxygen2:
https://cs.github.com/r-lib/roxygen2/blob/39d2dbefddac508dba6bffdec3a625cac7af16bc/R/util-locale.R#L7
* Rely on vctrs, which is already in `Imports`. This is what I actually did. At this moment, C locale ordering is only exposed in `vctrs::vec_rank()`, so I used that and set vctrs minimum version accordingly. In the future, we can switch to `vec_sort_radix()`.

---

If you check out readr from current HEAD (a257bfb at the time I write this), the files `R/cpp11.R` and `src/cpp11.cpp` are generated from a listing of the C++ files sorted according to `en_US.UTF-8`.

With CRAN decor and RStudio Version 2022.12.0+353 (2022.12.0+353):

* If you do `devtools::clean_dll()` and `cpp11::cpp_register()`, in the Console, `R/cpp11.R` and `src/cpp11.cpp` do not change. Still picking up the ambient locale.
* If you do `devtools::clean_dll()` and `devtools::document()`, in the Console, `R/cpp11.R` and `src/cpp11.cpp` do not change. Still picking up the ambient locale.
* If you then do `devtools::clean_dll()` and **use RStudio** to do `devtools::document()` (from *Build* pane or keyboard shortcut), `R/cpp11.R` and `src/cpp11.cpp` will change. It appears that RStudio silently changes to the C locale during the execution of `document()`.
  - I don't see the same locale-switching behaviour if I use the IDE to `test()` or `check()`. Only `document()`.

After this PR, anyone using cpp11 might see one instance of churn, getting their `R/cpp11.R` and `src/cpp11.cpp` to be based on C-locale-ordering of source files, but then they should only change for legit reasons after that.

---

Added later: this is where the IDE is doing this. Probably not ideal behaviour in the current era, but possibly also not worth changing.

https://github.com/rstudio/rstudio/blob/86ae4e281b82ad87ddce3247aab01144efdb6ecb/src/cpp/session/modules/build/SessionBuild.cpp#L544-L552

### CODE REVIEW

File: R/decor.R [hunk 1]
```diff
@@ -22,11 +22,15 @@ cpp_files <- function(pkg = ".") {
...
 
   src <- file.path(pkg, "src")
-  if (dir.exists(src)) {
-    return(list.files(src, full.names = TRUE, pattern = "[.](cc|cpp|h|hpp)$"))
+  if (!dir.exists(src)) {
+    return(character())
   }
 
-  return(character())
+  out <- list.files(src, full.names = TRUE, pattern = "[.](cc|cpp|h|hpp)$")
```
Comment (romainfrancois) [inline R/decor.R:31]: Is this something `fs::` could help with ? 

What's preventing to use `vec_sort()` ? cc @DavisVaughan @lionel- 

`decor` is mostly (only?) used at development time, so we can be easily depend on recent `vctrs`.

File: R/decor.R [hunk 2]
```diff
@@ -22,11 +24,15 @@ cpp_files <- function(pkg = ".") {
...
+  # always sort these paths according to the C locale to avoid nuisance changes
+  # in files generated downstream
+  # TODO: switch to vctrs::vec_sort_radix() or vctrs::vec_sort() when possible
+  out[order(vctrs::vec_rank(out))]
```
Comment (DavisVaughan) [inline R/decor.R:35]: I also happen to know we've used this in the past for similar reasons

```r
> roxygen2:::sort_c
function (x, ...) 
with_collate("C", sort(x, ...))
```

File: R/decor.R (hunk 1 above)
Comment (DavisVaughan) [reply R/decor.R:31]: `vec_order()` and `vec_sort()` don't use the C locale yet but will in a future vctrs release

**Comment by jennybc** [comment]
Folks already reviewed this while in draft form, but I have added the planned description now.

**Comment by jennybc** [comment]
@romainfrancois I think you can review this for real now

File: R/decor.R (hunk 1 above)
Comment (romainfrancois) [reply R/decor.R:31]: Thanks. We can revisit at this point then.

**Review by romainfrancois — APPROVED** [review]

**Pull request merged**